### ISSUE TITLE
Intermittent - unable to create lock in backend: circuit breaker open for file <lock/XXXXXXXXXX>

### ISSUE THREAD

**Issue by heapdavid** [restic/restic#5005]
Output of `restic version`
--------------------------
restic 0.17.0 compiled with go1.22.5 on linux/amd64

What backend/service did you use to store the repository?
---------------------------------------------------------
Backblaze B2

Problem description / Steps to reproduce
----------------------------------------

Hi,

We upload lots of large files to B2 on a nightly basis. Since upgrading to 0.17.0 on Debian, we're seeing occasional errors uploading to B2:

```
unable to create lock in backend: circuit breaker open for file <lock/c498f551d0>
bytes to write=10240, bytes written=4096, total written         60K
buffer (writer): write of data failed: Success
Unable to write to output file ()
```

We pipe tarballs into restic using the following command inside a script:

`... | restic backup --quiet --stdin --stdin-filename "${TARBALL}.tar" -o b2.connections=8`

If we set `RESTIC_FEATURES=backend-error-redesign=false` as mentioned in another issue then we don't see this issue - uploads happen reliably again.

Expected behavior
-----------------

Upload to B2 reliably

Actual behavior
---------------

As described above, 

Do you have any idea what may have caused this?
-----------------------------------------------
Given that setting `RESTIC_FEATURES=backend-error-redesign=false` works around this issue, it's likely to be something in that redesign.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Restic is a handy part of our backup regime, and has been reliable and straightforward to use. Thank you!

**Comment by MichaelEischer** [comment]
> Unable to write to output file ()

The error messages seem to be somewhat mangled. This error does not exist in restic.

> unable to create lock in backend: circuit breaker open for file <lock/c498f551d0>

Did restic log any errors before that (to stderr)? It looks like restic failed to access some lock files. I'll need those errors to know what the underlying problem is. The relevant log entries are probably prefixed with `Load(` and contain ` failed: `.

> We pipe tarballs into restic using the following command inside a script:
> 
> `... | restic backup --quiet --stdin --stdin-filename "${TARBALL}.tar" -o b2.connections=8`

Unrelated to the issue here: you might want to take a look at the new --stdin-from-command option, see https://restic.readthedocs.io/en/latest/040_backup.html#reading-data-from-a-command .

**Comment by heapdavid** [comment]
> The error messages seem to be somewhat mangled. This error does not exist in restic.

Apologies, looks like I've included error output from the script too. 

Checking back, we see 2 entries in the couple of seconds leading up to the `unable to create lock in backend` error

```
2024-08-13 12:11:08.181 Load(<lock/4b79b36dfc>, 0, 0) failed: b2_download_file_by_name: 404: File with such name does not exist.
2024-08-13 12:11:08.197 Load(<lock/4b79b36dfc>, 0, 0) failed: b2_download_file_by_name: 404: File with such name does not exist.
2024-08-13 12:11:09.908 unable to create lock in backend: circuit breaker open for file <lock/c498f551d0>
```

This is the same each time it fails, although the lock ids are different, e.g.

```
2024-08-13 13:43:08.154 Load(<lock/2e62c4eddd>, 0, 0) failed: b2_download_file_by_name: 404: File with such name does not exist.
2024-08-13 13:43:08.181 Load(<lock/2e62c4eddd>, 0, 0) failed: b2_download_file_by_name: 404: File with such name does not exist.
2024-08-13 13:43:09.819 unable to create lock in backend: circuit breaker open for file <lock/07a248fd7a>
```

---

> you might want to take a look at the new --stdin-from-command option

Thanks, looks like that might a cleaner way to do what we're doing 👍

**Comment by MichaelEischer** [comment]
> circuit breaker open for file <lock/c498f551d0>

There should also be errors regarding the specific lock file about which the circuit breaker complained. Are those also 404 errors? It would be helpful to get all errors printed by restic from the start until the locking failure.

> 2024-08-13 12:11:08.197 Load(<lock/4b79b36dfc>, 0, 0) failed: b2_download_file_by_name: 404: File with such name does not exist.

Those errors are somewhat unexpected. B2 is supposed to return a consistent file listing, so there should be at most one of those 404 errors per file.

I assume there are similar lock loading errors when `RESTIC_FEATURES=backend-error-redesign=false` is set? How do those look like?

How many locks does `restic list locks` return?

**Comment by heapdavid** [comment]
I managed to reproduce directly from the command line, changing restic to `--verbose` mode - this is all we get:

```
# echo "this is a test file, please delete" | restic backup --verbose --stdin --stdin-filename "dh-test-upload-file.txt" -o b2.connections=8
open repository
repository d8b8f43f opened (version 1)
Load(<lock/91b9421fff>, 0, 0) failed: b2_download_file_by_name: 404: File with such name does not exist.
Load(<lock/91b9421fff>, 0, 0) failed: b2_download_file_by_name: 404: File with such name does not exist.
unable to create lock in backend: circuit breaker open for file <lock/798ffec919>
```

> How many locks does `restic list locks` return?

We are using restic in parallel so wondering if that's the issue? We had 6 different restic processes running at the time so this list of locks looks correct:
```
# restic list locks
repository d8b8f43f opened (version 1)
3e9764438154e7721fc26d04c7a2bae93b35b32e944eb32c66858f59a8be111b
797a1fd96d9f1542704a4c8e6a270d5c2d2563c8a88fb450d7463dd7969d5c76
85f082637780083215b0c54ccf7a42cfa0dbfe4a157776f257b8f88214bfb6b0
8cdc13cdc1872ee7f4678d0c664ba101e89968fbb6cf6711f2c99d534f60f4ed
c0d8ed00a5c7944dd959c14d859ae7a58d36fc30be676c42ad7b993b5c34ec7e
e63be718f5b497177b7778c3d553328c2386acee3e7cc1a38846a7f479370360
```
> I assume there are similar lock loading errors when `RESTIC_FEATURES=backend-error-redesign=false` is set? How do those look like?

Our script is still running with `RESTIC_FEATURES=backend-error-redesign=false` set - Looks like we do see similar 404s logged occasionally, with a backoff of some kind for retrying:

```
Aug 21 10:30:02 <hostname> <scriptname>.sh[2117307]: Load(<lock/1f47f96fc2>, 0, 0) returned error, retrying after 337.359251ms: b2_download_file_by_name: 404: File with such name does not exist.
Aug 21 10:30:02 <hostname> <scriptname>.sh[2117307]: Load(<lock/1f47f96fc2>, 0, 0) returned error, retrying after 762.884398ms: b2_download_file_by_name: 404: File with such name does not exist.
Aug 21 10:30:03 <hostname> <scriptname>.sh[2117307]: Load(<lock/1f47f96fc2>, 0, 0) returned error, retrying after 1.065339388s: b2_download_file_by_name: 404: File with such name does not exist.
Aug 21 10:30:04 <hostname> <scriptname>.sh[2117307]: Load(<lock/1f47f96fc2>, 0, 0) returned error, retrying after 2.238756821s: b2_download_file_by_name: 404: File with such name does not exist.
Aug 21 10:30:06 <hostname> <scriptname>.sh[2117307]: Load(<lock/1f47f96fc2>, 0, 0) returned error, retrying after 1.64823168s: b2_download_file_by_name: 404: File with such name does not exist.
Aug 21 10:30:08 <hostname> <scriptname>.sh[2117307]: Load(<lock/1f47f96fc2>, 0, 0) returned error, retrying after 5.668982297s: b2_download_file_by_name: 404: File with such name does not exist.
Aug 21 10:30:13 <hostname> <scriptname>.sh[2117307]: Load(<lock/1f47f96fc2>, 0, 0) returned error, retrying after 2.893698713s: b2_download_file_by_name: 404: File with such name does not exist.
Aug 21 10:30:16 <hostname> <scriptname>.sh[2117307]: Load(<lock/1f47f96fc2>, 0, 0) returned error, retrying after 6.415179783s: b2_download_file_by_name: 404: File with such name does not exist.
Aug 21 10:30:23 <hostname> <scriptname>.sh[2117307]: Load(<lock/1f47f96fc2>, 0, 0) returned error, retrying after 6.453273439s: b2_download_file_by_name: 404: File with such name does not exist.
Aug 21 10:30:29 <hostname> <scriptname>.sh[2117307]: Load(<lock/1f47f96fc2>, 0, 0) returned error, retrying after 11.120150638s: b2_download_file_by_name: 404: File with such name does not exist.
Aug 21 10:30:41 <hostname> <scriptname>.sh[2117307]: Load(<lock/1f47f96fc2>, 0, 0) failed: b2_download_file_by_name: 404: File with such name does not exist.
```

**Comment by MichaelEischer** [comment]
> Load(<lock/91b9421fff>, 0, 0) failed: b2_download_file_by_name: 404: File with such name does not exist.
> unable to create lock in backend: circuit breaker open for file <lock/798ffec919>

Thanks for testing. After a bit of digging I found a bug in the circuit breaker logic. The bug could cause the circuit breaker for unrelated lock files to trip when some lock file failed to load, see https://github.com/restic/restic/pull/5011 .

However, based on the symptoms seen so far, this will just change the error message towards
```
unable to create lock in backend: b2_download_file_by_name: 404: File with such name does not exist.
```

> We are using restic in parallel so wondering if that's the issue? We had 6 different restic processes running at the time so this list of locks looks correct:

Having multiple running instances is necessary to trigger the circuit breaker bug.

> Looks like we do see similar 404s logged occasionally, with a backoff of some kind for retrying:

Does restic manage to lock the repository in those cases? Or is the "failed: b2_download_file_by_name: 404: File with such name does not exist." error logged multiple times?

**Comment by heapdavid** [comment]
Great, thanks for looking into this!

> Does restic manage to lock the repository in those cases? Or is the "failed: b2_download_file_by_name: 404: File with such name does not exist." error logged multiple times?

Seems like it does manage to get a lock and continue successfully, the `failed: b2_download_file_by_name: 404: File with such name does not exist.` error is not logged again for that particular process.

**Comment by MichaelEischer** [comment]
> Seems like it does manage to get a lock and continue successfully

Looks like it can take a few seconds until the file listing no longer contains removed files. I've added short delays between the locking retries in https://github.com/restic/restic/pull/5012 . These should be enough to mimic the delay introduced by the old retry logic (`backend-error-redesign=false`).

Would it be possible to give the code from the PR a try?

**Comment by heapdavid** [comment]
Apologies, I was on leave and didn't see this until now. Thanks for merging the fix, I've updated to the new release without the workaround, will let you know if we see any further issues.
